cpp/tests/experimental/trajectory/trajectory_test_utils.cuh

```diff
@@ -0,0 +1,192 @@
+
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/gather.h>
+#include <thrust/random.h>
+#include <thrust/sequence.h>
+#include <thrust/shuffle.h>
+#include <thrust/tabulate.h>
+#include <thrust/transform.h>
+
+#include <chrono>
+#include <cstdint>
+
+namespace cuspatial {
+namespace test {
+
+using time_point = std::chrono::time_point<std::chrono::system_clock, std::chrono::milliseconds>;
+
+/* Test data generation for trajectory APIs. Generates num_trajectories trajectories of random
+   size between 1 and max_trajectory_size samples. Creates both a reference (sorted) set of
+   trajectory IDs, timestamps, and points, and a shuffled (unsorted) set of the same for test
+   input. The sorted data can be input directly into `trajectory_distance_and_speed`, while
+   the shuffled data can be input to `derive_trajectories`.
+
+   The times are not random, but do vary somewhat between trajectories and trajectory lengths.
+   Each trajectory has a distinct starting time point and trajectory timestamps increase
+   monotonically at each sample. The interval between samples varies within a small range.
+
+   Likewise, the positions are not random, but follow a sinusoid pattern based on the time stamps.
+*/
+template <typename T>
+struct trajectory_test_data {
+  rmm::device_vector<std::int32_t> offsets;
+
+  rmm::device_vector<std::int32_t> ids;
+  rmm::device_vector<time_point> times;
+  rmm::device_vector<cuspatial::vec_2d<T>> points;
+
+  rmm::device_vector<std::int32_t> ids_sorted;
+  rmm::device_vector<time_point> times_sorted;
+  rmm::device_vector<cuspatial::vec_2d<T>> points_sorted;
+
+  trajectory_test_data(std::size_t num_trajectories, std::size_t max_trajectory_size)
+  {
+    make_data(num_trajectories, max_trajectory_size);
+  }
```

Comment:
This function doesn't seem to contribute anything except to add one more level of indirection?

Reply:
This is a constructor for the class.  I generally don't like putting a ton of code inside a constructor, and `make_data` is a long function.